src/native/eventpipe/ep-provider.c

```diff
@@ -78,6 +78,11 @@ provider_prepare_callback_data (
 	EP_ASSERT (provider != NULL);
 	EP_ASSERT (provider_callback_data != NULL);
 
+	ep_requires_lock_held ();
+
+	if (provider->callback_func != NULL)
```

Comment:
Any reason why this can't go inside ep_provider_callback_data_init and then do the decrement and potential signaling in ep_provider_callback_data_fini when callback data gets freed?

Reply:
I'd be fine with that but I'm curious, do you want to just get rid of provider_prepare_callback_data and the caller should directly invoke provider_callback_data_init? It doesn't seem like prepare has a reason to exist if we'll move all the logic into init.

Reply:
I guess the same point applies here then, since we need to hold lock when working with the provider, we don't want to add locking into the init function since it can be called with or without holding the lock (since its mainly setting up the internals of the callback data).

Reply:
yeah, if we don't want init/fini to use any locking (which feels like a reasonable invariant to me) then how it is now seems good to me. We still only have one place we are incrementing and one place we are decrementing. If there is ever a need in the future to have more than one place we could do some refactoring at that time.

Reply:
Agree, my feeling is that all this should probably happen closer to the callback invocation and not when preparing the callback data and then decrement after the running the callback since those two places can be spread out over time and code location, but as long its only in one place, we could hold that refactoring.